query-service-impl/src/main/java/org/hypertrace/core/query/service/pinot/QueryRequestToPinotSQLConverter.java

```diff
@@ -449,4 +450,25 @@ private String convertLiteralToString(LiteralConstant literal, Params.Builder pa
     }
     return ret;
   }
+
+  private boolean isColumnMapAttribute(Expression expression) {
+    return expression.getValueCase() == COLUMNIDENTIFIER
+        && isMapField(expression.getColumnIdentifier().getColumnName());
+  }
+
+  private boolean isAttributeExpressionMapAttribute(Expression expression) {
```

Comment:
Here, we're giving a different meaning to mapAttribute than right above. Above, we're considering an expression a map attribute if it simply references a map-typed attributes (e.g. "tags"). Here it needs to both do that and subpath into it.

This means that if I just selected `tags` via attribute expression, code like `convertExpressionToMapKeyColumn` would error out.

Based on how things are being used, I think I'd suggest reforming these four into three functins:
- `isMapAttributeExpression` (true if either a column identifier references a map or attribute expression references a map, regardless of subpath)
- `isMapSubpathAttributeExpression` (true for attribute expression w/ subpath)
- `isSimpleAttributeExpression` (true if either a column identifier or attribute expression without a subpath.

I am calling everything Attribute Expression here regardless of col identifier vs attribute expression with the assumption that col identifier will be removed shortly. If they were going to coexist, it would make sense to clarify in the predicate names, perhaps by changing "AttributeExpression" to something like "AttributeReference" to highlight the abstraction.

Reply:
>This means that if I just selected tags via attribute expression, code like convertExpressionToMapKeyColumn would error out.

convertExpressionToMapKeyColumn is not getting called from any such instance. Agreed there is some confusion but that is because we support both columnIdentifier and attributeExpression now which would be deprecated later.

Reply:
> convertExpressionToMapKeyColumn is not getting called from any such instance.

You're correct that it's not being called on selection (although the ambiguity there is the whole problem - it requires understanding the implementation of these functions because their names don't match their behavior). Where it could be called though, would be the CONTAINS_KEY case. I left a comment on the new util ( https://github.com/hypertrace/query-service/pull/107/files#r764918332 ) if we wrote that with attribute expressions, I'm pretty sure it would cause an error. 

> Agreed there is some confusion but that is because we support both columnIdentifier and attributeExpression now which would be deprecated later.

This is a simple fix now, and makes it easier to do that deprecation later since it helps encapsulate the distinction between the two. 

-  `isMapSubpathAttributeExpression` is just a rename of `isAttributeExpressionMapAttribute`. This is behavior we'd want to persist regardless of deprecation.
-  `isMapAttributeExpression` replaces everywhere we do `isColumnMapAttribute(expression) || isAttributeExpressionMapAttribute(expression)` - this both makes the later deprecation easier and fixes the bug discussed above
- `isSimpleAttributeExpression` is just a rename of `isColumnExpression` - again, relevant regardless of deprecation.

Reply:
resolved